File: 3/go/part1/part1.go
package main

import (
	"log"
	"os"
	"regexp"
	"strconv"
)

func main() {
	if len(os.Args) < 2 {
		log.Fatal("provide part and input file")
	}
	inputFile := os.Args[1]

	input, err := os.ReadFile(inputFile)
	if err != nil {
		log.Fatal(err)
	}

	m, _ := regexp.Compile(`mul\([0-9]+,[0-9]+\)`)
	d, _ := regexp.Compile(`[0-9]+`)

	muls := m.FindAllString(string(input), -1)

	total := 0

	for _, mul := range muls {
		digits := d.FindAllString(string(mul), -1)
		first, err := strconv.Atoi(digits[0])
		if err != nil {
			log.Fatal(err)
		}
		second, err := strconv.Atoi(digits[1])
		if err != nil {
			log.Fatal(err)
		}
		total += first * second
	}

	log.Println(total)
}

File: 3/py/part1.py
import sys
import string

if len(sys.argv) < 2:
    print("provide input file")

inputfile = sys.argv[1]
input = ""

with open(inputfile, "r") as f:
    input = f.read()

current_string = ""
first_value = ""
second_value = ""
total = 0

for char in input:
    
    if char == "m":
        current_string = "m"
    elif char == "u" and current_string == "m":
        current_string = "mu"
    elif char == "l" and current_string == "mu":
        current_string = "mul"
    elif char == "(" and current_string == "mul":
        current_string = "mul("
        first_value = ""
        second_value = ""
    elif char in string.digits:
        if current_string == "mul(":
            first_value += char
        elif current_string == "mul(,":
            second_value += char
    elif char == "," and current_string == "mul(":
        current_string = "mul(,"
    elif char == ")" and current_string == "mul(,":
        try:
            current_string = ""
            total += int(first_value)*int(second_value)
        except:
            pass
    else:
        current_string = ""

print(total)